hail/src/main/scala/is/hail/expr/ir/lowering/LowerTableIR.scala

```diff
@@ -20,95 +19,138 @@ object TableStage {
+    case ((name, value), body) => Let(name, value, body)
   }
 
-  def mapPartition(f: IR => IR): TableStage = {
-    val outer = this
-    new TableStage(letBindings, broadcastVals, globals, partitioner, contexts) {
-      def partition(ctxRef: Ref): IR = f(outer.partition(ctxRef))
-    }
-  }
+  def mapPartition(f: IR => IR): TableStage =
+    copy(partitionIR = f(partitionIR))
 
-  def mapPartitionWithContext(f: (IR, Ref) => IR): TableStage = {
-    val outer = this
-    new TableStage(letBindings, broadcastVals, globals, partitioner, contexts) {
-      def partition(ctxRef: Ref): IR = f(outer.partition(ctxRef), ctxRef)
-    }
-  }
+  def mapPartitionWithContext(f: (IR, Ref) => IR): TableStage =
+    copy(partitionIR = f(partitionIR, Ref(ctxRefName, ctxType)))
 
   def mapContexts(f: IR => IR)(getOldContext: IR => IR): TableStage = {
-    val outer = this
-    new TableStage(letBindings, broadcastVals, globals, partitioner, f(contexts)) {
-      override def partition(ctxRef: Ref): IR = bindIR(getOldContext(ctxRef))(outer.partition)
-    }
+    val newContexts = f(contexts)
+    TableStage(letBindings, broadcastVals, globals, partitioner, newContexts, ctxRef => bindIR(getOldContext(ctxRef))(partition))
   }
 
   def mapGlobals(f: IR => IR): TableStage = {
     val newGlobals = f(globals)
-    val newID = genUID()
-    val outer = this
-    new TableStage(
-      letBindings :+ newID -> newGlobals,
-      broadcastVals + newID,
-      Ref(newID, newGlobals.typ),
-      partitioner, contexts) {
-      def partition(ctxRef: Ref): IR = outer.partition(ctxRef)
-    }
+    val globalsRef = Ref(genUID(), newGlobals.typ)
+
+    copy(
+      letBindings = letBindings :+ globalsRef.name -> newGlobals,
+      broadcastVals = broadcastVals :+ globalsRef.name -> globalsRef,
+      globals = globalsRef)
   }
 
-  def collect(bind: Boolean = true): IR = {
-    val ctx = Ref(genUID(), types.coerce[TStream](contexts.typ).elementType)
-    val broadcastRefs = MakeStruct(letBindings.filter { case (name, _) => broadcastVals.contains(name) })
-    val glob = Ref(genUID(), broadcastRefs.typ)
-    val cda = CollectDistributedArray(contexts, broadcastRefs,
-      ctx.name, glob.name,
-      broadcastVals.foldLeft(partition(ctx))((accum, name) => Let(name, GetField(glob, name), accum)))
-    if (bind) wrapInBindings(cda) else cda
+  def mapCollect(f: IR => IR): IR = {
+    mapCollectWithGlobals(f) { (parts, globals) => parts }
   }
 
-  def collectWithGlobals(): IR = {
-    val mapped = mapPartition(ToArray)
-    mapped.wrapInBindings(
-      MakeStruct(FastIndexedSeq(
-        "rows" -> ToArray(flatMapIR(ToStream(mapped.collect(bind = false))) { elt => ToStream(elt) }),
-        "global" -> mapped.globals)))
+  def mapCollectWithGlobals(mapF: IR => IR)(body: (IR, IR) => IR): IR =
+    mapCollectWithContextsAndGlobals((part, ctx) => mapF(part))(body)
+
+  def mapCollectWithContextsAndGlobals(mapF: (IR, Ref) => IR)(body: (IR, IR) => IR): IR = {
```

Comment:
I had to add this during the rebase, to support the lowered TableWriter without using `wrapInBindings`.